src/lib/storage/dictionary_segment/dictionary_segment_iterable.hpp

```diff
@@ -90,42 +108,51 @@ class DictionarySegmentIterable : public PointAccessibleSegmentIterable<Dictiona
       _chunk_offset += n;
     }
 
-    bool equal(const Iterator& other) const { return _attribute_it == other._attribute_it; }
+    bool equal(const Iterator& other) const { return _chunk_offset == other._chunk_offset; }
 
-    std::ptrdiff_t distance_to(const Iterator& other) const { return other._attribute_it - _attribute_it; }
+    std::ptrdiff_t distance_to(const Iterator& other) const {
+      return static_cast<std::ptrdiff_t>(other._chunk_offset) - _chunk_offset;
+    }
 
     SegmentPosition<T> dereference() const {
       const auto value_id = static_cast<ValueID>(*_attribute_it);
-      const auto is_null = (value_id == _null_value_id);
 
-      if (is_null) return SegmentPosition<T>{T{}, true, _chunk_offset};
-
-      return SegmentPosition<T>{T{*(_dictionary_begin_it + value_id)}, false, _chunk_offset};
+      if constexpr (IsNullable) {
+        if (value_id == _null_value_id) {
+          return SegmentPosition<T>{T{}, true, _chunk_offset};
+        }
+      }
+
+      if constexpr (std::is_same_v<Dictionary, FixedStringVector>) {
+        return SegmentPosition<T>{_dictionary->get_string_at(value_id), false, _chunk_offset};
```

Comment:
Return a NonNullSegmentPosition here?

Reply:
(Of course we need to make the return type dependent on the type of the iterator)

Reply:
Thaaaat's exactly what was bothering me. We know we return a false for NULL, but still check `position.is_null()` everytime. If that helps, that would be super awesome. Will test it.